File: clients/python_pypy/model/action.py
from model.action_type import ActionType
from stream_wrapper import StreamWrapper

class Action:
    """Action unit is currently performing"""

    __slots__ = ("finish_tick","action_type",)

    finish_tick: int
    action_type: ActionType

    def __init__(self, finish_tick: int, action_type: ActionType):
        self.finish_tick = finish_tick
        """Tick when the action will be finished"""
        self.action_type = action_type
        """Type of the action"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Action":
        """Read Action from input stream
        """
        finish_tick = stream.read_int()
        action_type = ActionType(stream.read_int())
        return Action(finish_tick, action_type)
    
    def write_to(self, stream: StreamWrapper):
        """Write Action to output stream
        """
        stream.write_int(self.finish_tick)
        stream.write_int(self.action_type)
    
    def __repr__(self):
        return "Action(" + \
            repr(self.finish_tick) + \
            ", " + \
            repr(self.action_type) + \
            ")"

File: clients/cpp/model/Action.cpp
#include "Action.hpp"

namespace model {

Action::Action(int finishTick, model::ActionType actionType) : finishTick(finishTick), actionType(actionType) { }

// Read Action from input stream
Action Action::readFrom(InputStream& stream) {
    int finishTick = stream.readInt();
    model::ActionType actionType = readActionType(stream);
    return Action(finishTick, actionType);
}

// Write Action to output stream
void Action::writeTo(OutputStream& stream) const {
    stream.write(finishTick);
    stream.write((int)(actionType));
}

// Get string representation of Action
std::string Action::toString() const {
    std::stringstream ss;
    ss << "Action { ";
    ss << "finishTick: ";
    ss << finishTick;
    ss << ", ";
    ss << "actionType: ";
    ss << actionTypeToString(actionType);
    ss << " }";
    return ss.str();
}

bool Action::operator ==(const Action& other) const {
    return finishTick == other.finishTick && actionType == other.actionType;
}

}

size_t std::hash<model::Action>::operator ()(const model::Action& value) const {
    size_t result = 0;
    result ^= std::hash<int>{}(value.finishTick) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<model::ActionType>{}(value.actionType) + 0x9e3779b9 + (result << 6) + (result >> 2);
    return result;
}